records/2022-11-04.reliability.sms-limits.md

```diff
@@ -0,0 +1,58 @@
+# SMS Limits
+
+Date: 2022-11-04
+
+## Status
+
+**DRAFT**
+
+## Context
+
+Originally Notify only had a "combined" email and SMS limit. 
+
+These proposed changes were prompted by an incident where we had a very high and unexpected volume of SMS sends.
+
+## Implementation overview
+
+In order to enforce a new SMS-specific limit, we have added checks to the "front door" of the API, where the post requests come in that result in an SMS send:
+- for the `/<uuid:service_id>/send-notification` endpoint used by the Admin "send one off" flow: [app/service/send_notification.py](https://github.com/cds-snc/notification-api/blob/d3412ad3697162550d3f5d73958f488a305d3cf7/app/service/send_notification.py#L66)
+- for the job creation endpoint used by the Admin CSV send flow: [app/job/rest.py](https://github.com/cds-snc/notification-api/blob/d3412ad3697162550d3f5d73958f488a305d3cf7/app/job/rest.py#L149)
+- for the `/bulk` API endpoint defined in [app/v2/notifications/post_notifications.py](https://github.com/cds-snc/notification-api/blob/d3412ad3697162550d3f5d73958f488a305d3cf7/app/v2/notifications/post_notifications.py#L186)
+- for the `/sms` API endpoint defined in [app/v2/notifications/post_notifications.py](https://github.com/cds-snc/notification-api/blob/d3412ad3697162550d3f5d73958f488a305d3cf7/app/v2/notifications/post_notifications.py#L228)
+
+Before arriving at this solution, we tried adding checks to other places in the API code (ie celery), including where the notifications are created in the database and where they are sent to SES. These other locations proved prone to errors and difficult to test, which led us to move all related code to the "front door".
```

Comment:
```suggestion
Before arriving at this solution, we tried adding checks to other places in the API code (ie celery), including where the notifications are created in the database and where they are sent to SES. These other locations proved prone to errors and difficult to test, which led us to move all related code to the "front door".

Placing the limits later in the data pipeline is alluring if we want a precise limit to be honored by the system and a single point to apply the logic for the logic. The inconvenience though is that as the data pipeline might process notifications in minutes, the API front door still continue to process intake messages. There is a time window when the system accepted more messages than the limits should have accepted, as it is at the sending stage. This has the side effect of having idle notifications caught in the system. This presents additional challenges of what to do with these notifications. Do we wait for the next day when the limit will lower again or should we mark these notifications as failures, unbeknownst to the sender unless they continually check the status of the notification? 

Considering this might be a bad user experience, which could also lead to additional design challenge, moving the logic to the front door to inform the user as early as we can seemed the best option.
```